core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -214,67 +322,82 @@ private void checkNotClosed() {
   }
 
   /**
-   * Reads and delivers as many messages to the sink as possible.
+   * Notify the listener that more messages may be available for consumption. This is called
+   * by both the transport and application threads.
    */
-  private void deliver() {
-    // We can have reentrancy here when using a direct executor, triggered by calls to
-    // request more messages. This is safe as we simply loop until pendingDelivers = 0
-    if (inDelivery) {
-      return;
-    }
-    inDelivery = true;
-    try {
-      // Process the uncompressed bytes.
-      boolean stalled = false;
-      while (pendingDeliveries > 0 && readRequiredBytes()) {
-        switch (state) {
-          case HEADER:
-            processHeader();
-            break;
-          case BODY:
-            // Read the body and deliver the message.
-            processBody();
-
-            // Since we've delivered a message, decrement the number of pending
-            // deliveries remaining.
-            pendingDeliveries--;
-            break;
-          default:
-            throw new AssertionError("Invalid state: " + state);
+  private void notifyListener() {
+    Iterator<InputStream> messages = new Iterator<InputStream>() {
+      Frame current;
+
+      @Override
+      public InputStream next() {
+        if (!hasNext()) {
+          throw new IllegalStateException("No element");
+        }
+        try {
+          return current.stream;
+        } finally {
+          current = null;
         }
       }
-      // We are stalled when there are no more bytes to process. This allows delivering errors as
-      // soon as the buffered input has been consumed, independent of whether the application
-      // has requested another message.
-      stalled = !isDataAvailable();
-
-      if (endOfStream) {
-        if (!isDataAvailable()) {
-          listener.endOfStream();
-        } else if (stalled) {
-          // We've received the entire stream and have data available but we don't have
-          // enough to read the next frame ... this is bad.
-          throw Status.INTERNAL.withDescription("Encountered end-of-stream mid-frame")
-              .asRuntimeException();
+
+      @Override
+      public boolean hasNext() {
+        int consumedBytes = 0;
+        Frame peek;
+        try {
+          if (current != null) {
+            return true;
+          }
+          do {
+            peek = frameQueue.peek();
+            if (peek == null) {
+              return false;
+            }
+            if (peek == END_OF_STREAM) {
+              // Transport enqueued EOS, so empty queue and report to the listener
+              frameQueue.remove();
```

Comment:
Is it possible that to iterators are called concurrently, resulting in `listener.endOfStream` called twice?